src/main/java/org/webeid/security/validator/validators/SubjectCertificateNotRevokedValidator.java

```diff
@@ -23,88 +23,175 @@
-
-            final CertificateStatus status = first.getCertStatus();
-
-            if (status instanceof RevokedStatus) {
-                RevokedStatus revokedStatus = (RevokedStatus) status;
-                throw (revokedStatus.hasRevocationReason() ?
-                    new UserCertificateRevokedException("Revocation reason: " + revokedStatus.getRevocationReason()) :
-                    new UserCertificateRevokedException());
-            } else if (status instanceof UnknownStatus) {
-                throw new UserCertificateRevokedException("Unknown status");
-            } else if (status == null) {
-                LOG.debug("OCSP check result is GOOD");
-            } else {
-                throw new UserCertificateRevokedException("Status is neither good, revoked nor unknown");
+            verifyOcspResponse(basicResponse, ocspService, certificateId);
+            if (!ocspService.isNonceDisabled()) {
+                checkNonce(request, basicResponse);
             }
-        } catch (CertificateEncodingException | OCSPException | IOException e) {
+        } catch (OCSPException | CertificateException | OperatorCreationException | IOException e) {
             throw new UserCertificateRevocationCheckFailedException(e);
         }
     }
+
+    private void verifyOcspResponse(BasicOCSPResp basicResponse, OcspService ocspService, CertificateID requestCertificateId) throws UserCertificateRevokedException, UserCertificateRevocationCheckFailedException, OCSPException, CertificateException, OperatorCreationException {
+        // As we sent the request for only a single certificate, we expect only a single response.
+        if (basicResponse.getResponses().length != 1) {
+            throw new UserCertificateRevocationCheckFailedException("OCSP response must contain one response, "
+                + "received " + basicResponse.getResponses().length + " responses instead");
+        }
+        // We require the responder to include a single certificate in the certs field of the response
+        // that helps us to verify the responder's signature.
+        if (basicResponse.getCerts().length != 1) {
+            throw new UserCertificateRevocationCheckFailedException("OCSP response must contain one responder certificate, "
+                + "received " + basicResponse.getResponses().length + " certificates instead");
+        }
+        final SingleResp certStatusResponse = basicResponse.getResponses()[0];
+        if (certStatusResponse.getCertID() != requestCertificateId) {
+            throw new UserCertificateRevocationCheckFailedException("OCSP responded with certificate ID that differs from the requested ID");
+        }
+
+        final X509CertificateHolder responderCert = basicResponse.getCerts()[0];
+
+        ocspService.validateResponderCertificate(responderCert, basicResponse.getProducedAt());
+        validateResponseSignature(basicResponse, responderCert);
+        validateSubjectCertificateStatus(certStatusResponse);
```

Comment:
Shouldn't we validate status before everything else?

Reply:
This would pose a security risk. The attacker would be able to set up a malicious OCSP responder that returns revoked statuses for valid certificates in this case, causing a denial of service.